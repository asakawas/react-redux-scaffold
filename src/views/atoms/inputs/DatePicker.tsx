import React, { FunctionComponent, useState } from 'react';
import {
  KeyboardDatePicker as MuiKeyboardDatePicker,
  KeyboardDatePickerProps as MuiKeyboardDatePickerProps,
} from '@material-ui/pickers';
import { MaterialUiPickersDate } from '@material-ui/pickers/typings/date';
import moment from 'moment';

/**
 * The type of props of DatePicker.
 */
export type DatePickerProps = Readonly<
  Required<Pick<MuiKeyboardDatePickerProps, 'name' | 'label'>> &
    Pick<MuiKeyboardDatePickerProps, 'inputRef'> & {
      /** If true or omitted, this input become read-only. */
      readOnly?: boolean;

      /** The default value (e.g. 1983-12-14). */
      defaultValue?: moment.MomentInput;
    }
>;

const DatePicker: FunctionComponent<DatePickerProps> = ({
  name,
  label,
  inputRef,
  readOnly,
  defaultValue,
}) => {
  const [selectedDate, setSelectedDate] = useState<MaterialUiPickersDate>(
    defaultValue ? moment(defaultValue) : null,
  );

  return (
    <MuiKeyboardDatePicker
      name={name}
      id={name}
      label={label}
      InputProps={{
        readOnly,
        inputProps: {
          'aria-label': `input-${name}`,
        },
      }}
      readOnly={readOnly}
      inputRef={inputRef}
      helperText={readOnly ? 'Read-only' : undefined}
      value={selectedDate}
      onChange={(date: MaterialUiPickersDate) => setSelectedDate(date)}
      format="YYYY-MM-DD"
      KeyboardButtonProps={readOnly ? { disabled: true } : undefined}
    />
  );
};

export default React.memo(DatePicker);
