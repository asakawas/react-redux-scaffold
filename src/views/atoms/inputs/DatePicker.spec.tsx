import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import MomentUtils from '@date-io/moment';
import DatePicker from './DatePicker';

describe('DatePicker', () => {
  test('is rendered as expected', () => {
    const { asFragment } = render(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <DatePicker name="test" label="Test" />
      </MuiPickersUtilsProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('readOnly ver is rendered as expected', () => {
    const { asFragment } = render(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <DatePicker name="test" label="Test" readOnly />
      </MuiPickersUtilsProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('has an empty value if defaultValue is omitted"', () => {
    const renderResult = render(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <DatePicker name="test" label="Test" />
      </MuiPickersUtilsProvider>,
    );

    const tree = renderResult.getByLabelText('input-test');
    // @ts-ignore  Property 'value' does not exist on type 'HTMLElement'
    expect(tree.value).toBe('');
  });

  test('has an empty value if defaultValue is empty"', () => {
    const renderResult = render(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <DatePicker name="test" label="Test" defaultValue="" />
      </MuiPickersUtilsProvider>,
    );

    const tree = renderResult.getByLabelText('input-test');
    // @ts-ignore  Property 'value' does not exist on type 'HTMLElement'
    expect(tree.value).toBe('');
  });

  test('has the given default value in the format of "YYYY-MM-DD"', () => {
    const renderResult = render(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <DatePicker name="test" label="Test" defaultValue="1234-01-31" />
      </MuiPickersUtilsProvider>,
    );

    const tree = renderResult.getByLabelText('input-test');
    // @ts-ignore  Property 'value' does not exist on type 'HTMLElement'
    expect(tree.value).toBe('1234-01-31');
  });

  test('can be typed in if readOnly is false', async () => {
    const renderResult = render(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <DatePicker name="test" label="Test" />
      </MuiPickersUtilsProvider>,
    );

    const tree = renderResult.getByLabelText('input-test');
    await userEvent.type(tree, '19831214');
    // @ts-ignore  Property 'value' does not exist on type 'HTMLElement'
    expect(tree.value).toBe('1983-12-14');
  });

  test('can not be typed in if readOnly is true', async () => {
    const renderResult = render(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <DatePicker name="test" label="Test" readOnly />
      </MuiPickersUtilsProvider>,
    );

    const tree = renderResult.getByLabelText('input-test');
    await userEvent.type(tree, '19831214');
    // @ts-ignore  Property 'value' does not exist on type 'HTMLElement'
    expect(tree.value).toBe('');
  });

  test("doesn't has a helper text to show 'Read-only' if readOnly is false", () => {
    const renderResult = render(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <DatePicker name="test" label="Test" />
      </MuiPickersUtilsProvider>,
    );

    const tree = renderResult.queryByText('Read-only');
    expect(tree).toBeNull();
  });

  test('has a helper text to show "Read-only" if readOnly is true', () => {
    const renderResult = render(
      <MuiPickersUtilsProvider utils={MomentUtils}>
        <DatePicker name="test" label="Test" readOnly />
      </MuiPickersUtilsProvider>,
    );

    const tree = renderResult.queryByText('Read-only');
    expect(tree).not.toBeNull();
  });
});
