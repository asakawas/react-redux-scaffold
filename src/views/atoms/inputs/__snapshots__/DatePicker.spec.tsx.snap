// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker is rendered as expected 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink="false"
      for="test"
      id="test-label"
    >
      Test
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
    >
      <input
        aria-invalid="false"
        aria-label="input-test"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
        id="test"
        name="test"
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DatePicker readOnly ver is rendered as expected 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root"
    readonly=""
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink="false"
      for="test"
      id="test-label"
    >
      Test
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
    >
      <input
        aria-describedby="test-helper-text"
        aria-invalid="false"
        aria-label="input-test"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
        id="test"
        name="test"
        readonly=""
        type="text"
        value=""
      />
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
    <p
      class="MuiFormHelperText-root"
      id="test-helper-text"
    >
      Read-only
    </p>
  </div>
</DocumentFragment>
`;
